### PR TITLE
search: auto-escape special regex runes if they would not give useful results unescaped

### LINKED ISSUE #2125: Auto-escape/fixup queries that resemble common literal queries but are invalid regexps
As a user, if I perform certain common kinds of search queries (listed below), Sourcegraph tries parsing the queries as regexps, which fails. My intent is clear, and Sourcegraph should handle these cases better, as described  below.

- `myfuncname(` to `myfuncname\(`
- `myarray[` to `myarray\[`

Note that these fixups ONLY occur in unambiguous cases, where the query would fail to be parsed as a regexp and therefore return no results (just an error). We do not (in this issue) plan to do any auto-fixup of *valid* queries.

In case we missed any common cases, inspect the Sourcegraph.com search logs for regexp parse errors and add other cases to handle to this issue.

This is a partial and simpler fix for the problem described in #633.

### PR DESCRIPTION
**Pull request by ijt**
Fixes #2125.

### CODE REVIEW

**Comment by ijt** [comment]
PTAL

**Comment by ijt** [comment]
Just a minute. I'll fix the broken test...

**Comment by sqs** [comment]
@ijt See https://github.com/sourcegraph/sourcegraph/issues/2125#issuecomment-477640893. By limiting to only the simple cases, it will be easier to write safe transformations. This also would address @keegancsmith's comment:

> Maybe we should just make the fixes very very simple and only apply to very specific looking cases.

**Comment by ijt** [comment]
> Instead of leaving inline comments, I reviewed this by pulling the PR and adding more test cases. I've pushed the broken test cases that I would expect to work. Feel free to revert the tests / etc if my feedback is invalid.

I agree about the idempotency test at least. Thanks for adding that.

> I think the approach of just always apply certain syntax unaware transformations means we break legitimate regexs + make certain regexs impossible to write, unless the user writes an invalid version of it.

I don't know of legitimate regexes that would be broken, but if you can find some then please add tests for them and I'll see if I can fix them.

> The approach of only doing fixes if it fails to compile may be the only reasonable solution? Maybe we should just make the fixes very very simple and only apply to very specific looking cases.

I thought so too, but that doesn't handle the case of `()`, a valid but useless regex. It's way more useful to consider that as a literal string.

**Comment by ijt** [comment]
@sqs, I think you're right that we should go with the less risky subset for now. 

For future reference, I pushed fixes to make the more extensive autoFix pass @keegancsmith's tests and also a quickcheck randomized test. I'll push commits to remove support for the additional cases, but we have the commits there if we want to bring them back later.

> @ijt See [#2125 (comment)](https://github.com/sourcegraph/sourcegraph/issues/2125#issuecomment-477640893). By limiting to only the simple cases, it will be easier to write safe transformations. This also would address @keegancsmith's comment:
> 
> > Maybe we should just make the fixes very very simple and only apply to very specific looking cases.

**Comment by ijt** [comment]
Okay, I pushed the more limited version. PTAL.

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,8 @@ All notable changes to Sourcegraph are documented in this file.
 
 ### Added
 
+- In search queries, some characters that are normally treated specially within regular expressions are automatically
```
Comment (nicksnyder) [inline CHANGELOG.md]: I think one specific example in the changelog would be appropriate.

**Pull request merged**